### PR TITLE
SONARPY-696 Typeshed: Support importing of submodules

### PR DESCRIPTION
**Pull request by dmitry-golovinov-sonarsource**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java [hunk 1]
```diff
@@ -77,7 +77,7 @@ public static String fullyQualifiedModuleName(String packageName, String fileNam
     String moduleName = extensionIndex > 0
       ? fileName.substring(0, extensionIndex)
       : fileName;
-    if (moduleName.equals("__init__")) {
+    if (moduleName.equals("__init__") || packageName.endsWith("." + moduleName) || packageName.equals(moduleName)) {
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java:80]: I don't fully understand the reason behind those extra conditions.
I see that you want to simplify logic inside `SymbolTableBuilder` and `Typeshed`, however this is not what I would expect from this function, i.e. if I call `fullyQualifiedModuleName('mod', 'mod.py')` I would expect the fully qualified name to be `mod.mod`

Here you have an example in python repl:
https://repl.it/repls/LastJauntyGlueware

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 2]
```diff
@@ -178,22 +178,19 @@ public static Symbol symbolWithFQN(String stdLibModuleName, String fullyQualifie
...
     }
-    String packageName = "";
-    String fileName = moduleName;
-    if (splittedName.length != 1) {
-      packageName = String.join(".", Arrays.copyOf(splittedName, splittedName.length - 1));
-      fileName = splittedName[splittedName.length - 1];
-    }
-    return new ModuleDescription(resource, fileName, packageName);
+    String fileName = Paths.get(fullPathToModule).getFileName().toString();
+    fileName = fileName.substring(0, fileName.length() - 4);
```
[the remark this answers is not in the document]
Comment (dmitry-golovinov-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:192]: My bad. This was no longer needed

File: python-frontend/src/test/java/org/sonar/python/semantic/FullyQualifiedNameTest.java [hunk 3]
```diff
@@ -432,7 +432,7 @@ public void init_module_relative_import() {
     PythonVisitorContext context = new PythonVisitorContext(fileInput, pythonFile, null, "foo.bar");
     fileInput = context.rootTree();
     CallExpression callExpression = (CallExpression) getAllDescendant(fileInput, tree -> tree.is(Tree.Kind.CALL_EXPR)).get(0);
-    assertThat(callExpression.calleeSymbol().fullyQualifiedName()).isEqualTo("foo.fn");
+    assertThat(callExpression.calleeSymbol().fullyQualifiedName()).isEqualTo("fn");
```
[the remark this answers is not in the document]
Comment (dmitry-golovinov-sonarsource) [reply python-frontend/src/test/java/org/sonar/python/semantic/FullyQualifiedNameTest.java:435]: Actually, the whole example seems to be quite incorrect for me. 
https://repl.it/repls/ShortCooperativeMineral

File: python-frontend/src/test/java/org/sonar/python/semantic/SymbolUtilsTest.java [hunk 4]
```diff
@@ -232,6 +232,16 @@ public void package_name_by_file() {
     assertThat(pythonPackageName(new File(baseDir, "packages/sound/formats/wavread.py"), baseDir)).isEqualTo("sound.formats");
   }
 
+  @Test
+  public void fqn_by_package_with_subpackage() {
+    assertThat(SymbolUtils.fullyQualifiedModuleName("", "foo.py")).isEqualTo("foo");
+    assertThat(SymbolUtils.fullyQualifiedModuleName("foo", "__init__.py")).isEqualTo("foo");
+    assertThat(SymbolUtils.fullyQualifiedModuleName("foo", "foo.py")).isEqualTo("foo");
```
[the remark this answers is not in the document]
Comment (dmitry-golovinov-sonarsource) [reply python-frontend/src/test/java/org/sonar/python/semantic/SymbolUtilsTest.java:239]: Agreed

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java (hunk 1 above)
Comment (dmitry-golovinov-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java:80]: You are right. Got rid of this extra.

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 5]
```diff
@@ -173,23 +176,33 @@ public static Symbol symbolWithFQN(String stdLibModuleName, String fullyQualifie
...
-  private static InputStream getResourceForModule(String moduleName, String categoryPath) {
-    InputStream resource = TypeShed.class.getResourceAsStream(categoryPath + moduleName + ".pyi");
-    if (resource != null) {
-      return resource;
+  @Nullable
+  private static ModuleDescription getResourceForModule(String moduleName, String categoryPath) {
+    String[] moduleNameHierarchy = moduleName.split("\\.");
+    String pathToModule = String.join("/", moduleNameHierarchy);
+    String moduleFileName = moduleNameHierarchy[moduleNameHierarchy.length - 1];
+    String targetModuleName = String.join(".", Arrays.copyOfRange(moduleNameHierarchy, 0, moduleNameHierarchy.length - 1));
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:184]: I would suggest renaming it `packageName`

**Pull request merged**